### ISSUE TITLE
Explorer AutoEncoder from tensorflow

### ISSUE THREAD

**Issue by cokelaer** [Poulpikan/stageM1#9]
AutoEncoder to reduce the local noise. See if signal emerges around centromers.

- standarize the code 
- Build AutoEncoder
- Optimize
- plot error ( convergence model)
- predict

**Comment by cokelaer** [comment]
import tensorflow as tf
        from tensorflow.keras import layers, models, optimizers, losses
        import numpy as np
        import matplotlib.pyplot as plt
        from tqdm import tqdm
        
        # Hyperparameters
        nfeat = 2  # Number of encoded features
        lr = 0.001
        epochs = 50
        quiet = False
        
        # Assume data is a NumPy array
        input_dim = data.shape[1]
        
        # Build the autoencoder
        class Autoencoder(tf.keras.Model):
            def __init__(self, input_dim):
                super(Autoencoder, self).__init__()
                self.encoder = models.Sequential([
                    layers.Dense(nfeat, activation='relu')
                ])
                self.decoder = models.Sequential([
                    layers.Dense(input_dim, activation='sigmoid')
                ])
        
            def call(self, x):
                encoded = self.encoder(x)
                decoded = self.decoder(encoded)
                return decoded
        
        # Initialize model
        model = Autoencoder(input_dim)
        model.build(input_shape=(None, input_dim))
        optimizer = optimizers.Adam(learning_rate=lr)
        mse_loss = losses.MeanSquaredError()
        
        # Convert data to tensor
        data_tensor = tf.convert_to_tensor(data, dtype=tf.float32)
        
        # Training loop
        losses_list = []
        for epoch in tqdm(range(epochs), disable=quiet):
            with tf.GradientTape() as tape:
                output = model(data_tensor)
                loss = mse_loss(data_tensor, output)
            gradients = tape.gradient(loss, model.trainable_variables)
            optimizer.apply_gradients(zip(gradients, model.trainable_variables))
            losses_list.append(loss.numpy())
        
            if not quiet:
                print(f"Epoch {epoch+1}, Loss: {loss.numpy()}")
        
        # Plot losses
        plt.figure(2)
        plt.plot(losses_list)
        plt.xlabel('Epoch')
        plt.ylabel('Loss')
        plt.title('Training Loss')
        plt.show()
        
        # Compute reconstruction error
        reconstructed = model(data_tensor)
        errors = tf.reduce_mean(tf.math.squared_difference(data_tensor, reconstructed), axis=1)

**Comment by cokelaer** [comment]
Normalisation is important, 

    mean = np.mean(data, axis=0)
    std = np.std(data, axis=0) +1.e-8  # add epsilon to avoid divisoion by zero
    data = (data - mean) / std


for the peak detection, you may try:

                peaks, properties = find_peaks(errors, width=5, height=height, prominence=2, distance=20)

**Comment by cokelaer** [comment]
Fait
